src/backend/opencl/kernel/hamming.cl

```diff
@@ -0,0 +1,327 @@
+
+#ifdef USE_LOCAL_MEM
+    if (valid_feat) {
+        // Copy blockDim.x training features to shared memory
+        #pragma unroll
+        for (unsigned i = 0; i < FEAT_LEN; i++) {
+            l_train[i * get_local_size(0) + tid] = train[i * ntrain + f];
+        }
+    }
+    barrier(CLK_LOCAL_MEM_FENCE);
+#endif
+
+    for (int j = 0; j < (int)nquery; j++) {
+        l_dist[tid] = max_dist;
+
+        // Load one query feature that will be tested against all training
+        // features in current block
+        if (tid < FEAT_LEN && valid_feat) {
+            l_query[tid] = query[tid * nquery + j];
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+
+        unsigned dist = 0;
+        if (valid_feat) {
+            #pragma unroll
+            for (int k = 0; k < (int)FEAT_LEN; k++) {
+                // Calculate Hamming distance for 32-bits of descriptor and
+                // accumulates to dist
+#ifdef USE_LOCAL_MEM
+                dist += popcount(l_train[k * get_local_size(0) + tid] ^ l_query[k]);
+#else
+                dist += popcount(train[k * ntrain + f] ^ l_query[k]);
+#endif
+            }
+        }
+
+        // Only stores the feature index and distance if it's smaller
+        // than the best match found so far
+        if (valid_feat) {
+            l_dist[tid] = dist;
+            l_idx[tid]  = f;
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+
+        // Find best match in training features from block to the current
+        // query feature
+        if (tid < 128) {
+            if (l_dist[tid + 128] < l_dist[tid]) {
+                l_dist[tid] = l_dist[tid + 128];
+                l_idx[tid]  = l_idx[tid + 128];
+            }
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+        if (tid < 64) {
+            if (l_dist[tid + 64] < l_dist[tid]) {
+                l_dist[tid] = l_dist[tid + 64];
+                l_idx[tid]  = l_idx[tid + 64];
+            }
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+        if (tid < 32) {
+            if (l_dist[tid + 32] < l_dist[tid]) {
+                l_dist[tid] = l_dist[tid + 32];
+                l_idx[tid]  = l_idx[tid + 32];
+            }
+            if (l_dist[tid + 16] < l_dist[tid]) {
+                l_dist[tid] = l_dist[tid + 16];
+                l_idx[tid]  = l_idx[tid + 16];
+            }
+            if (l_dist[tid + 8] < l_dist[tid]) {
+                l_dist[tid] = l_dist[tid + 8];
+                l_idx[tid]  = l_idx[tid + 8];
+            }
+            if (l_dist[tid + 4] < l_dist[tid]) {
+                l_dist[tid] = l_dist[tid + 4];
+                l_idx[tid]  = l_idx[tid + 4];
+            }
+            if (l_dist[tid + 2] < l_dist[tid]) {
+                l_dist[tid] = l_dist[tid + 2];
+                l_idx[tid]  = l_idx[tid + 2];
+            }
+            if (l_dist[tid + 1] < l_dist[tid]) {
+                l_dist[tid] = l_dist[tid + 1];
+                l_idx[tid]  = l_idx[tid + 1];
+            }
+        }
```

Comment:
You can't rely on intra warp synchronization on OpenCL. Different devices have different criterion. Always use barriers.